[SPARK-22897][CORE]: Expose stageAttemptId in TaskContext

advancedxy opened the pull request:
## What changes were proposed in this pull request?
stageAttemptId added in TaskContext and corresponding construction modification

## How was this patch tested?
Added a new test in TaskContextSuite, two cases are tested:
1. Normal case without failure
2. Exception case with resubmitted stages

Link to [SPARK-22897](https://issues.apache.org/jira/browse/SPARK-22897)

advancedxy committed:
[SPARK-22897][CORE]: Expose stageAttemptId in TaskContext

advancedxy commented:
cc @cloud-fan

AmplabJenkins commented:
Can one of the admins verify this patch?

cloud-fan commented on core/src/main/scala/org/apache/spark/TaskContext.scala:
```diff
@@ -150,6 +150,10 @@ abstract class TaskContext extends Serializable {
   def stageId(): Int
 
+  /**
+   * The attempt ID of the stage that this task belongs to.
```
can we follow `taskAttemptId` and say more in the document?

advancedxy replied:
Of course, I can add more doc.

But is it necessary? I think the field(function) name is already self-explanatory.

jiangxb1987 commented:
Please explain in the JIRA/PR description about your use case that need to access the stageAttemptId in the executor - we must provide a very good reason to make such a change to a public interface.

cloud-fan commented:
ok to test

SparkQA commented:
**[Test build #85409 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85409/testReport)** for PR 20082 at commit [`5753ee0`](https://github.com/apache/spark/commit/5753ee0d938b507c61c19233f87043396befadc5).

SparkQA commented:
**[Test build #85409 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85409/testReport)** for PR 20082 at commit [`5753ee0`](https://github.com/apache/spark/commit/5753ee0d938b507c61c19233f87043396befadc5).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85409/
Test FAILed.

advancedxy commented:
>Please explain in the JIRA/PR description about your use case that need to access the stageAttemptId in the executor - we must provide a very good reason to make such a change to a public interface.

@jiangxb1987  quoted from  SPARK-22897:
>Currently, there's no easy way for Executor to detect a new stage is launched as stageAttemptId is missing.

In one of our use case, spark task needs to reset some internal state once the stage is changed(we don't want to reset the state for every task as there may be a lot of tasks).

IMHO, stageAttemptId is as important as stageId or taskAttemptId and should be provided to user.

As for compatibility, we should do our best to avoid any incompatibility in minor releases. However,
this change should be source and binary compatible with previous releases.

advancedxy committed:
Minor fixes such as:

1. Update mima exclude
2. Update wording in TaskContext
3. Update JavaTaskContextCompileCheck as suggested in TaskContext note

SparkQA commented:
**[Test build #85410 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85410/testReport)** for PR 20082 at commit [`f02bc1e`](https://github.com/apache/spark/commit/f02bc1eefb9d4cb9c47d0ccbb2dda2ef9f8a6b39).

advancedxy commented on project/MimaExcludes.scala:
```diff
@@ -95,7 +95,10 @@ object MimaExcludes {
+    ProblemFilters.exclude[FinalClassProblem]("org.apache.spark.ml.tuning.TrainValidationSplitModel$TrainValidationSplitModelWriter"),
+
+    // [SPARK-22897] Expose stageAttemptId in TaskContext
+    ProblemFilters.exclude[ReversedMissingMethodProblem]("org.apache.spark.TaskContext.stageAttemptId")
```
This change is suggested by https://github.com/apache/spark/pull/12248/

If not appropriate, please let me know.

cloud-fan replied:
can you put it at the beginning of `v23excludes`?

SparkQA commented:
**[Test build #85410 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85410/testReport)** for PR 20082 at commit [`f02bc1e`](https://github.com/apache/spark/commit/f02bc1eefb9d4cb9c47d0ccbb2dda2ef9f8a6b39).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85410/
Test FAILed.

advancedxy committed:
Update MimaExcludes to move stageAttemptId filter to the beginning

SparkQA commented:
**[Test build #85420 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85420/testReport)** for PR 20082 at commit [`0f86e95`](https://github.com/apache/spark/commit/0f86e95fd9c8703516e23b5b3bd370a89b27bc69).

SparkQA commented:
**[Test build #85420 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85420/testReport)** for PR 20082 at commit [`0f86e95`](https://github.com/apache/spark/commit/0f86e95fd9c8703516e23b5b3bd370a89b27bc69).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85420/
Test FAILed.

SparkQA commented:
**[Test build #85421 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85421/testReport)** for PR 20082 at commit [`59e4a9c`](https://github.com/apache/spark/commit/59e4a9c70c037729f3eb60b47b2e625208687385).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85421/
Test PASSed.

jiangxb1987 commented on core/src/test/scala/org/apache/spark/scheduler/TaskContextSuite.scala:
```diff
@@ -158,6 +159,28 @@ class TaskContextSuite extends SparkFunSuite with BeforeAndAfter with LocalSpark
+
+    // Check stage attemptIds that are resubmitted when task fails
+    val stageAttemptIdsWithFailedStage =
+      sc.parallelize(Seq(1, 2, 3, 4), 4).repartition(1).mapPartitions { _ =>
```
You don't need `repartition` here, just `sc.parallelize(Seq(1, 2, 3, 4), 1).mapPartitions {...}`

jiangxb1987 commented on core/src/test/scala/org/apache/spark/scheduler/TaskContextSuite.scala:
```diff
@@ -158,6 +159,28 @@ class TaskContextSuite extends SparkFunSuite with BeforeAndAfter with LocalSpark
+      sc.parallelize(Seq(1, 2, 3, 4), 4).repartition(1).mapPartitions { _ =>
+      val stageAttemptId = TaskContext.get().stageAttemptId()
+      if (stageAttemptId < 2) {
+        throw new FetchFailedException(null, 0, 0, 0, "Fake")
```
Emmm... just throw an `Exception` is enough here?

advancedxy replied:
Related to repartition part.

I use FetchFailedException to explicitly trigger a stage resubmission.  Otherwise, the task would be resubmitted in the same stage if IIRC.

jiangxb1987 replied:
oh, right~

jiangxb1987 replied:
Please add comment to explain that `FetchFailedException` will trigger a new stage attempt, while a common `Exception` will only trigger a task retry.

advancedxy replied:
Will do.

advancedxy committed:
Add comment for usage of FetchFailedException

SparkQA commented:
**[Test build #85429 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85429/testReport)** for PR 20082 at commit [`291bbbc`](https://github.com/apache/spark/commit/291bbbc8dbd82dd971727af8479ae7a6cfe5df96).

SparkQA commented:
**[Test build #85429 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85429/testReport)** for PR 20082 at commit [`291bbbc`](https://github.com/apache/spark/commit/291bbbc8dbd82dd971727af8479ae7a6cfe5df96).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

zsxwing commented on core/src/main/scala/org/apache/spark/TaskContextImpl.scala:
```diff
@@ -42,6 +42,7 @@ import org.apache.spark.util._
  */
 private[spark] class TaskContextImpl(
     val stageId: Int,
+    val stageAttemptId: Int,
```
nit: add `override`. Since you are touching this file, could you also add `override` to `stageId` and `partitionId`.

advancedxy replied:
Will do.

Would you tell me the difference or rationale?

cloud-fan replied:
it's kind of a code style standard: add `override` if it is override.

advancedxy replied:
OK then.

zsxwing reviewed (commented):
LGTM except one nit

cloud-fan commented on core/src/main/scala/org/apache/spark/TaskContext.scala:
```diff
@@ -150,6 +150,11 @@ abstract class TaskContext extends Serializable {
+  /**
+   * An ID that is unique to the stage attempt that this task belongs to.
+   */
+  def stageAttemptId(): Int
```
I think we should call it `stageAttempNumber` to be consistent with `taskAttemptNumber`. Also let's follow the comment of `attemptNumber`

advancedxy replied:
Yeah, if we are defining `stageAttemptId` from scratch, I would go for `stageAttemptNumber`.  However `stageAttemptId` are already used elsewhere in the codebase, Like in [Task.scala](https://github.com/apache/spark/blob/ded6d27e4eb02e4530015a95794e6ed0586faaa7/core/src/main/scala/org/apache/spark/scheduler/Task.scala#L55). I think it's more important to be consistent.

However I could update the comment to reflect the attempt number part if you wish

advancedxy committed:
Add override for stageId, stageAttemptId and partitionId.

Updating wording in TaskContext.stageAttemptId

SparkQA commented:
**[Test build #85457 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85457/testReport)** for PR 20082 at commit [`72a3abf`](https://github.com/apache/spark/commit/72a3abf8110a13c3719b8d6a600edee509b36ae9).

SparkQA commented:
**[Test build #85457 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85457/testReport)** for PR 20082 at commit [`72a3abf`](https://github.com/apache/spark/commit/72a3abf8110a13c3719b8d6a600edee509b36ae9).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85457/
Test FAILed.

jiangxb1987 commented:
retest this please

SparkQA commented:
**[Test build #85474 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85474/testReport)** for PR 20082 at commit [`72a3abf`](https://github.com/apache/spark/commit/72a3abf8110a13c3719b8d6a600edee509b36ae9).

SparkQA commented:
**[Test build #85474 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85474/testReport)** for PR 20082 at commit [`72a3abf`](https://github.com/apache/spark/commit/72a3abf8110a13c3719b8d6a600edee509b36ae9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85474/
Test PASSed.

advancedxy commented:
ping @cloud-fan @jiangxb1987 @zsxwing, I think it's ready for merging.

cloud-fan commented on core/src/main/scala/org/apache/spark/TaskContext.scala:
```diff
@@ -150,6 +150,13 @@ abstract class TaskContext extends Serializable {
+   * times the stage has been attempted. The first stage attempt will be assigned stageAttemptId = 0
+   * , and subsequent attempts will increasing stageAttemptId one by one.
+   */
+  def stageAttemptId(): Int
```
My concern is that, internally we use `stageAttemptId`, and internally we call `TaskContext.taskAttemptId` `taskId`. However, for end users, they don't know the internal code, and they are more familiar with `TaskContext`. I think the naming should be consistent with the public API `TaskContext`, instead of internal code.

advancedxy replied:
I have no objection for either ‘id’ nor ‘number’, they are both reasonable. 

I am on train now。 If no other input, I can rename it to ‘stageAttemptNumber’ since you insisted.

advancedxy committed:
Rename stageAttemptId to stageAttemptNumber

SparkQA commented:
**[Test build #85512 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85512/testReport)** for PR 20082 at commit [`2581263`](https://github.com/apache/spark/commit/2581263dd99b7ab788b0412f8650ed9115ca6477).

SparkQA commented:
**[Test build #85513 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85513/testReport)** for PR 20082 at commit [`9266cd8`](https://github.com/apache/spark/commit/9266cd8d4558b675b081a7282c626d79bb6bb786).

SparkQA commented:
**[Test build #85512 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85512/testReport)** for PR 20082 at commit [`2581263`](https://github.com/apache/spark/commit/2581263dd99b7ab788b0412f8650ed9115ca6477).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85512/
Test PASSed.

SparkQA commented:
**[Test build #85513 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85513/testReport)** for PR 20082 at commit [`9266cd8`](https://github.com/apache/spark/commit/9266cd8d4558b675b081a7282c626d79bb6bb786).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85513/
Test PASSed.

advancedxy commented:
@cloud-fan Please take another look.

cloud-fan commented on core/src/main/scala/org/apache/spark/scheduler/Task.scala:
```diff
@@ -79,6 +79,7 @@ private[spark] abstract class Task[T](
     SparkEnv.get.blockManager.registerTask(taskAttemptId)
     context = new TaskContextImpl(
       stageId,
+      stageAttemptId, // stageAttemptId and stageAttemptNumber are semantically equal
```
How much work we need to rename the internal `stageAttemptId` to `stageAttemptNumber`?

advancedxy replied:
The modification may not be too much (100+ occurrences in 20+ files), however it may break eventLog's JsonProtocol backward compatibility(not sure)..

@squito you may have more knowledge on this since you introduced `stageAttemptId`.

cloud-fan replied:
ah, so `stageAttemptId` is already exposed in developer API, we can't change it.

cloud-fan commented:
LGTM

cloud-fan commented:
thanks, merging to master/2.3!

zsxwing commented:
@advancedxy could you also submit a PR to add a `def stageAttemptNumber = stageAttemptId` to `StageInfo` and mark `stageAttemptId` as deprecated? It will make our public APIs more consistent.

advancedxy commented:
Of course.

squito commented:
sorry this is a little late, but lgtm too.  agree with the points above about leaving the old name deprecated and moving to the new name
